### DISCUSSION TITLE
[Question] Unexpectedly Identical Training Results on Different PCs (Seed Fixed, torch_deterministic=False)

### DISCUSSION THREAD

**Discussion by jhs3330395** [isaac-sim/IsaacLab#4556] [Q&A]
- Isaac Lab Version: 2.3.1
- Isaac Sim Version: 5.1.0

I have a question. As far as I know, even if the seed is fixed, training on different PCs with different CPUs usually results in different learning outcomes.

Moreover, in Isaac Lab’s manager-based RL with RSL-RL, the seed is set using the **set_seed** function from **torch_utils**, and the default option is that **torch_deterministic** is set to **False**.

**However, when I trained a G1 humanoid walking robot on two different PCs, I obtained 100% identical logs and results.**

Can this be considered a normal situation, or does it indicate a potential problem?

Additionally, when I used net_forces_w as part of the reward or observation, the results from the two PCs diverged.
Surprisingly, using only net_forces_w_history did not cause any differences—the two PCs produced identical outputs as long as the seed was the same.

**Comment by RandomOakForest** [comment]
Thank you for posting this. Identical results across two PCs with the same seed can be normal in Isaac Lab, and the behavior you see with `net_forces_w` vs `net_forces_w_history` is actually a strong hint about where non‑determinism is coming from.[^1][^2][^3] I'll move this post to our _Discussions_ for follow up.

## Why identical training on two PCs can be “normal”

In Isaac Lab’s manager-based RL, the environment’s `seed()` ultimately calls `torch_utils.set_seed(seed)`, which seeds Python, NumPy, PyTorch, and Omniverse Replicator in a consistent way. If the two PCs have:[^5][^1]

- Same Isaac Lab version (2.3.1) and Isaac Sim (5.1.0).[^1]
- Same GPU model and driver / CUDA stack (or at least same backend used).
- Same RL code, configs, and number of envs / workers.

then it is entirely possible to get bit‑wise identical runs, even with `torch_deterministic=False` (which mainly controls whether certain slow deterministic kernels are forced, not whether “everything must be non‑deterministic”). In practice, many PyTorch and Isaac Lab operations are deterministic enough that two very similar systems can produce identical logs over long training horizons.[^6][^7]

So: identical logs across two machines does not by itself indicate a bug; it usually means your hardware + software stacks are matching closely enough for the operations you use.[^7][^6]

## Why `net_forces_w` causes divergence but `net_forces_w_history` does not

Isaac Lab’s contact sensor provides at least two key buffers:

- `net_forces_w`: current-step net contact force in world frame per body.[^2][^3]
- `net_forces_w_history`: a history buffer where each new step shifts older forces and inserts the new `net_forces_w` at index 0.[^2]

A typical update looks like (simplified):[^2]

```python
# net_forces_w updated from PhysX each step
# history update
net_forces_w_history[env_ids, 1:] = net_forces_w_history[env_ids, :-1].clone()
net_forces_w_history[env_ids, 0] = net_forces_w[env_ids]
```

Key points:

- `net_forces_w` depends directly on the physics contact resolution at that time step, and contact resolution is one of the most sensitive and potentially non‑deterministic parts of PhysX (tiny floating‑point differences or multithreading order can change which contacts are detected or how they are accumulated).[^3]
- Once a specific `net_forces_w` value is produced on a given step, the later history operations are **pure tensor copies** and indexing on the GPU in a fixed order, which are typically deterministic across runs that share the same execution configuration.[^2]

This fits your observation:

- When `net_forces_w` is part of the reward or observation: any small machine‑specific differences in contact forces feed directly into the policy inputs or rewards, get amplified over many steps, and cause divergence across PCs.
- When using only `net_forces_w_history`: you are still using past forces, but if the underlying physics and kernel scheduling happen to align exactly in your current setup, the whole pipeline (force computation + history shift + RL training) ends up deterministic between the two machines. The history operations themselves don’t introduce new non‑determinism.

In other words, `net_forces_w` is closer to the raw, potentially non‑deterministic physics output, while `net_forces_w_history` is just a deterministic temporal buffer on top of whatever forces PhysX produced.[^3][^2]

## Should you worry about identical runs?

Not necessarily. What you see is compatible with:

- “Accidental determinism”: your two PCs are similar enough that all relevant kernels and PhysX code paths behave identically for this particular task and configuration.
- “Conditional non‑determinism”: when you expose particularly sensitive signals (like raw contact forces) into the reward / observation, small machine differences show up; when you don’t, they don’t.

However, you should keep these in mind:

- **Reproducibility guarantee is limited**: both PyTorch and physics engines explicitly do not guarantee identical results across machines or versions, even with fixed seeds. You should not rely on cross‑machine bit‑exact reproducibility as a contractual property.[^6][^7]
- **Regression detection**: your current behavior (same logs on two PCs) is actually helpful; it means if results suddenly diverge after modifying code, you can treat that as a real regression signal, provided your stack stays the same.
- **Physics sensitivity**: features tied directly to contact forces (`net_forces_w`) are more likely to expose non‑determinism than smoother kinematic quantities like joint positions / velocities.


## Practical recommendations

To systematically reason about this in Isaac Lab 2.3.1 / Isaac Sim 5.1.0:[^5][^1][^3]

- Check that both machines have:
    - Same GPU architecture and driver + CUDA version.
    - Same Isaac Lab commit / wheel (not just version string).
    - Same number of CPU threads / `num_envs` / async workers.
- If you want **more determinism**, you can:
    - Pin PhysX and Isaac Sim settings (fixed substeps, disable adaptive features, fixed randomization schedule).
    - Consider turning on PyTorch deterministic algorithms where viable (e.g., `torch.use_deterministic_algorithms(True)`), understanding there is a performance trade‑off.[^6]
- If you want to **test sensitivity**:
    - Train twice on the *same* machine with same seed, but change `num_envs` or small parts of the reward.
    - Compare how much `net_forces_w` vs kinematic observations change outcome; this quantifies how fragile your reward design is to contact noise.

So, your findings are consistent with known behavior of Isaac Lab / PhysX / PyTorch: identical runs are not a red flag by themselves, and the divergence triggered specifically by `net_forces_w` reflects the higher sensitivity and potential non‑determinism of raw contact forces rather than an obvious bug in your setup.


[^1]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.envs.html

[^2]: https://github.com/isaac-sim/IsaacLab/discussions/2026

[^3]: https://isaac-sim.github.io/IsaacLab/main/source/overview/core-concepts/sensors/contact_sensor.html

[^5]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab_rl/isaaclab_rl.html

[^6]: https://stackoverflow.com/questions/67511658/training-pytorch-models-on-different-machines-leads-to-different-results

[^7]: https://github.com/pytorch/pytorch/issues/18412

[^8]: https://github.com/isaac-sim/IsaacLab/issues/4545

[^9]: https://docs.robotsfan.com/isaaclab_official/v2.0.0/_modules/isaaclab/envs/manager_based_env.html

[^10]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab_mimic/isaaclab_mimic.envs.html

[^11]: https://docs.robotsfan.com/isaaclab_official/main/_modules/isaaclab_rl/sb3.html

[^12]: https://lmsys.org/blog/2025-09-22-sglang-deterministic/

[^13]: https://www.reddit.com/r/pytorch/comments/fo23lt/cant_achive_reproducability_determinism_in/

[^14]: https://github.com/isaac-sim/IsaacLab/issues/1208

[^15]: https://www.youtube.com/watch?v=oss4n1NWDKo

**Reply by jhs3330395** [reply]
“Thank you very much. Your answers have been extremely clear and helpful. I have one more question.

I have been using Isaac Lab for a long time. 

When I was using Isaac Lab Version 2.1.0 with Isaac Sim Version 4.5.0,

I also used net_forces_w, but on the same PC with the same seed, the results were always identical.

This suggests that the deterministic behavior of net_forces_w might have changed as the Lab version was updated.

Does this mean that net_forces_w has become non‑deterministic in the newer versions? And if so, can we say that this non‑determinism is actually preferable?”
